pkg/bpfstats/bpfstats.go

```diff
@@ -0,0 +1,122 @@
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bpfstats
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"sync"
+
+	"github.com/cilium/ebpf"
+	"golang.org/x/sys/unix"
+)
+
+type BPFStatsMethod int
+
+const (
+	// MethodNone means that no call to EnableBPFStats() has been made or was unsuccessful
+	MethodNone BPFStatsMethod = iota
+
+	// MethodBPFFunc uses stats collection via BPF(BPF_ENABLE_STATS)
+	MethodBPFFunc
+
+	// MethodSysctl uses stats collection via sysctl (/proc/sys/kernel/bpf_stats_enabled)
+	MethodSysctl
+)
+
+var (
+	mutex     sync.Mutex
+	refCnt    int
+	statsSock io.Closer
+	method    = MethodNone
+)
+
+// EnableBPFStats enables collection of bpf program stats. It tries to use BPF_ENABLE_STATS first
+// (which requires Linux >= 5.8). If that fails, it will fall back to trying to
+// enable via sysctl (/proc/sys/kernel/bpf_stats_enabled). This function will make
+// sure that repeated calls will not enable stats collection more than once. Instead,
+// it will keep track of the number of calls and only stop stat collection, when
+// DisableBPFStats() has been called the same number of times.
+func EnableBPFStats() error {
+	mutex.Lock()
+	defer mutex.Unlock()
+
+	refCnt++
+
+	if refCnt != 1 {
+		return nil
+	}
+
+	// Actually enable
+	s, err := ebpf.EnableStats(unix.BPF_STATS_RUN_TIME)
+	if err != nil {
+		// Use fallback method
+		err = ioutil.WriteFile("/proc/sys/kernel/bpf_stats_enabled", []byte("1"), 0o644)
+		if err != nil {
+			refCnt = 0
+			return fmt.Errorf("could not enable stat collection: %w", err)
+		}
+		method = MethodSysctl
+	} else {
+		statsSock = s
+		method = MethodBPFFunc
+	}
+
+	return nil
+}
+
+// DisableBPFStats disables collection of bpf program stats if no consumer
+func DisableBPFStats() error {
+	mutex.Lock()
+	defer mutex.Unlock()
+
+	refCnt--
+
+	if refCnt < 0 {
+		refCnt = 0
+		return errors.New("bpf stat collection already disabled")
+	}
+
+	if refCnt != 0 {
+		return nil
+	}
+
+	// Actually disable
+	switch method {
+	case MethodBPFFunc:
+		err := statsSock.Close()
+		statsSock = nil
+		if err != nil {
+			return fmt.Errorf("could not disable stat collection using BPF(): %w", err)
+		}
+	case MethodSysctl:
+		err := ioutil.WriteFile("/proc/sys/kernel/bpf_stats_enabled", []byte("0"), 0o644)
+		if err != nil {
+			return fmt.Errorf("could not disable stat collection using sysctl: %w", err)
+		}
+	}
+
+	return nil
+}
+
+// GetMethod returns the currently used method to enable stats collection. If
+// EnableBPFStats() has not yet been called, it will return MethodNone.
+func GetMethod() BPFStatsMethod {
+	mutex.Lock()
+	defer mutex.Unlock()
+	return method
+}
```

Comment:
Is this function useful?

Reply:
Since it's a package that could be used from third parties as well, I thought it could be helpful to know which method is being used. I can remove it, if you don't think it is.

Reply:
We can leave it there, I just wanted to be sure that it wasn't a leftover.